### PR TITLE
Complete refactoring of variant and related annotation records

### LINKED ISSUE #96: Add back dbSnp ID field
See discussion on https://github.com/bigdatagenomics/adam/issues/1103

### PR DESCRIPTION
**Pull request by heuermh**
Fixes #96.

Still needs work; additional changes to come based on VCF mapping review.

### CODE REVIEW

File: src/main/resources/avro/bdg.avdl [hunk 1]
```diff
@@ -363,12 +363,55 @@ record NucleotideContigFragment {
...
+  /**
+   Breakend.  VCF INFO reserved key "SVTYPE" value "BND".
+   */
+  BREAKEND,
+
+  /**
+   Copy number variable region (may be both deletion and duplication).
+   VCF INFO reserved key "SVTYPE" value "CNV".
+   */
+  COPY_NUMBER_VARIABLE,
```
Comment (fnothaft) [inline src/main/resources/avro/bdg.avdl]: `s/variable/variant/g`? Or, is `variable` meant to indicate that the copy number of this region varies?

File: src/main/resources/avro/bdg.avdl [hunk 2]
```diff
@@ -411,34 +449,36 @@ record StructuralVariant {
...
+   Calculated by start + referenceAllele.length().
    */
   union { null, long } end = null;
 
   /**
-   A string describing the reference allele at this site.
+   Zero or more of unique names or identifiers for this variant. If this is a dbSNP variant it is
+   encouraged to use the rs number(s).  VCF column 3 "ID".
+   */
+  array<string> names = [];
```
Comment (fnothaft) [inline src/main/resources/avro/bdg.avdl]: OOC, any reason to pick `name` vs. `id`? I feel like ID would be (slightly) more consistent with other records.

**Comment by fnothaft** [comment]
LGTM other than two questions.

File: src/main/resources/avro/bdg.avdl (hunk 1 above)
Comment (heuermh) [inline src/main/resources/avro/bdg.avdl]: verbatim from spec, it sucks but ... ?

File: src/main/resources/avro/bdg.avdl (hunk 2 above)
Comment (heuermh) [inline src/main/resources/avro/bdg.avdl]: consistency with GA4GH; I don't feel strongly either way, `ids` or `identifiers` vs `names`

Comment (fnothaft) [inline src/main/resources/avro/bdg.avdl]: SGTM. Do we want to refactor throughout to be consistent? `AlignmentRecord` is name as well, `Feature` and `Sample` are ID?

File: src/main/resources/avro/bdg.avdl (hunk 1 above)
Comment (fnothaft) [inline src/main/resources/avro/bdg.avdl]: Actually, it's not too bad: the crux is "this region has varied copy number across samples". A "copy number variant" would be the specific change in copy number. I actually think this is the right call.

Comment (heuermh) [inline src/main/resources/avro/bdg.avdl]: I agree, though unfortunate TLA collision.

Comment (fnothaft) [inline src/main/resources/avro/bdg.avdl]: Yeah, no joke.

File: src/main/resources/avro/bdg.avdl (hunk 2 above)
Comment (heuermh) [inline src/main/resources/avro/bdg.avdl]: GA4GH use a strong definition of `id`
https://github.com/ga4gh/schemas/blob/master/doc/source/api/apidesign_intro.rst#object-ids
https://github.com/ga4gh/schemas/blob/master/doc/source/api/apidesign_intro.rst#id-and-name

I don't believe we care.  What are you proposing?

Comment (fnothaft) [inline src/main/resources/avro/bdg.avdl]: Yeah, actually, nevermind.

**Comment by jpdna** [comment]
> LGTM other than two questions.

What are the remaining two questions or other issues on this PR?

**Comment by fnothaft** [comment]
Thanks for pinging @jpdna. They were the two inline comments I added, but @heuermh answered the questions and I didn't think we needed any additional changes, so this is +1 from my end.

**Comment by jpdna** [comment]
Does the complementary code for the ADAM repo to populate the `array<string> names` in `Variant` exist now in any PR, or is that a task we need to to open an issue for?   
Even if that ADAM repo code is not ready yet - are ready to go ahead and merge this bdg-formats PR now in prep for that?

**Comment by fnothaft** [comment]
> Does the complementary code for the ADAM repo to populate the array<string> names in Variant exist now in any PR, or is that a task we need to to open an issue for?

I don't think the code is done yet. I think it might be added to https://github.com/bigdatagenomics/adam/pull/1078, but I'd wait for @heuermh to comment.

> Even if that ADAM repo code is not ready yet - are ready to go ahead and merge this bdg-formats PR now in prep for that?

+1

**Comment by jpdna** [comment]
So - what's the SOP on merging bdg-formats PRs?  I see we don't have a `scripts/commit-pr.sh` as we do for ADAM.

**Comment by fnothaft** [comment]
Yeah, we just hit the green merge button here. It's pretty rare that we:

1. have multiple active PRs in this repo
2. have PRs that need huge complex rebases

in this repo, so the `./scripts/commit-pr.sh` script isn't as necessary for bdg-formats.

**Comment by jpdna** [comment]
Thanks @heuermh 
merged

**Comment by heuermh** [comment]
No!  This wasn't ready to go.  Sorry for not replying sooner, have been in the mountains away from cell phone service.

**Comment by jpdna** [comment]
Ah sorry - what steps should we take to roll it back?  I see a handy "revert" button above in github that create a PR to revert.
Should I push the revert button?

**Comment by fnothaft** [comment]
No rush on this. We can either revert the merge, or force push HEAD prior to the merge to master. I'm fine either way. The first one is arguably the more kosher one to do.

**Comment by heuermh** [comment]
Being that there may be some concern in my ability to land all the relevant code changes before the 0.20 release of adam, please let me take the opportunity to consider rolling this and commit 5e81a9c0b073fe9163288d288621ee83af8b12a7 back, and re-proposing some of the changes piecemeal.

We hadn't finished going though all the potential variant and genotype schema changes before I went on vacation, and this pull request didn't even reflect the latest conversation we had (for example, I believe we decided to remove `StructuralVariant` entirely).

**Comment by fnothaft** [comment]
@heuermh you're in the office next week, right? If so, let's chat through this on standup.

**Comment by heuermh** [comment]
Back by Wed for sure.

**Comment by fnothaft** [comment]
Sounds good; we'll talk then, and I hope you're having a great trip!

**Pull request merged**